### PR TITLE
[msl-out] Make varyings' struct members unique

### LINKED ISSUE #1490: [metal-out] Duplicate member in input structs
I don't have access to any macos hardware, but one of my users reported this error:
```
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
    Internal error in VERTEX shader: Metal: Compilation failed: 

program_source:71:19: error: duplicate member 'position'
    metal::float2 position [[attribute(1)]];
                  ^
program_source:70:19: note: previous declaration is here
    metal::float2 position [[attribute(0)]];
                  ^
program_source:84:3: error: type 'vs_mainInput' is not valid for attribute 'stage_in'
  vs_mainInput varyings [[stage_in]]
  ^                       ~~~~
```

I'm suspecting that the shader that causes this looks like this:
```rust
struct Vertex {
    [[location(0)]] position: vec2<f32>;
};

struct NoteInstance{
    [[location(1)]] position: vec2<f32>;
    [[location(2)]] size: vec2<f32>;
    [[location(3)]] color: vec3<f32>;
    [[location(4)]] radius: f32;
};

struct VertexOutput {

};

[[stage(vertex)]]
fn vs_main(vertex: Vertex, note: NoteInstance) -> VertexOutput {
    var out: VertexOutput;
    return out;
}
```
Naga outputs this metal code for it:
```metal
struct vs_mainInput {
    metal::float2 position [[attribute(0)]];
    metal::float2 position [[attribute(1)]];
    metal::float2 size [[attribute(2)]];
    metal::float3 color [[attribute(3)]];
    float radius [[attribute(4)]];
};
```
As we can see there are two `position` fields in this output, and that causes an error

[full-shader-input-and-output.zip](https://github.com/gfx-rs/naga/files/7425749/full-shader-input-and-output.zip)

---
Kinda similar issue: https://github.com/gfx-rs/naga/issues/1464

### PR DESCRIPTION
**Pull request by evahop**
Fixes #1490

Naming conflicts can occur when multiple structs get flattened into one varyings struct.
This moves the `namer.call` invocation into the flattening routine so that each member gets a unique name.

### CODE REVIEW

**Review by teoxoy — CHANGES_REQUESTED** [review]
Thanks for the PR! It looks good but could you add a test for this? The one in the issue would work.

**Comment by evahop** [comment]
Yes, I was going to add that there's also a `namer.namespace` method specifically for structs, It may be more appropriate, but I am not familiar enough with result types. I will try to use that instead, but no promises.

**Comment by evahop** [comment]
Added tests, including one that confirms #1464 has also been fixed (not by me).
Abandoned  `namer.namespace` since `built-ins` don't necessarily end up in the struct, 
I just added a new `namer` solely for the locations.

**Comment by evahop** [comment]
Added snapshot, including one that confirms #1464 has also been fixed (not by me).
Abandoned  `namer.namespace` and just ended up just added a new `namer` solely for the locations. It causes a bit of duplication, since I also stick them in `flattened_member_names` if you want.

**Comment by evahop** [comment]
Added snapshot, which also confirms #1464 has been fixed (not by me).
Abandoned  `namer.namespace` and ended up just added a new `namer` solely for the locations. It causes a bit of duplication, since I also stick them in `flattened_member_names` if you want.

**Comment by evahop** [comment]
Added the snapshot, which also confirms #1464 has been fixed (not by me).
Abandoned  `namer.namespace` and ended up just adding a new `namer` solely for the locations.

File: src/back/msl/writer.rs [hunk 1]
```diff
@@ -3704,6 +3713,16 @@ impl<W: Write> Writer<W> {
...
+                            let name_key = &NameKey::StructMember(arg.ty, member_index);
+                            flattened_member_names.insert(
+                                NameKey::StructMember(arg.ty, member_index),
+                                match member.binding {
+                                    Some(crate::Binding::Location { .. }) => {
+                                        varyings_namer.call(&self.names[name_key])
+                                    }
+                                    _ => self.namer.call(&self.names[name_key]),
+                                },
+                            );
```
Comment (teoxoy) [inline src/back/msl/writer.rs:3725]:
```suggestion
                            let name_key = NameKey::StructMember(arg.ty, member_index);
                            let name = match member.binding {
                                Some(crate::Binding::Location { .. }) => {
                                    varyings_namer.call(&self.names[&name_key])
                                }
                                _ => self.namer.call(&self.names[&name_key]),
                            };
                            flattened_member_names.insert(name_key, name);
```

**Comment by teoxoy** [comment]
Seems like https://github.com/gfx-rs/naga/issues/1464 was fixed by https://github.com/gfx-rs/naga/pull/1766.

**Review by teoxoy — APPROVED** [review]
Good stuff!

**Pull request merged**